File: homework3/cpp/SecureStackTester.cpp
#include "SecureStackCpp.h"

#include <cstring>
#include <iostream>
#include <limits>
#include <memory>
#include <string>

namespace {

int tests_run = 0;
int tests_failed = 0;

void record_result(const bool condition, const char *test_name) {
    ++tests_run;
    if (condition) {
        std::cout << "[PASS] " << test_name << '\n';
        return;
    }

    ++tests_failed;
    std::cout << "[FAIL] " << test_name << '\n';
}

void test_create_and_destroy() {
    const StackCreateResultCpp create_result = stack_create_cpp(3U);
    StackBoolResultCpp empty_result{false, StackStatusCodeCpp::InvalidState, "", false};
    StackBoolResultCpp full_result{false, StackStatusCodeCpp::InvalidState, "", false};
    StackResultCpp destroy_result{false, StackStatusCodeCpp::InvalidState, ""};

    record_result(create_result.success && create_result.stack != nullptr,
                  "stack_create_cpp returns a success result and non-null stack");
    if (!create_result.success || create_result.stack == nullptr) {
        return;
    }

    empty_result = stack_is_empty_cpp(create_result.stack);
    full_result = stack_is_full_cpp(create_result.stack);

    record_result(empty_result.success && empty_result.value,
                  "stack_is_empty_cpp succeeds for a new stack");
    record_result(full_result.success && !full_result.value,
                  "stack_is_full_cpp succeeds for a new stack");

    destroy_result = stack_destroy_cpp(create_result.stack);
    record_result(destroy_result.success, "stack_destroy_cpp succeeds for a valid stack");

    destroy_result = stack_destroy_cpp(nullptr);
    record_result(!destroy_result.success && destroy_result.code == StackStatusCodeCpp::InvalidArgument,
                  "stack_destroy_cpp fails fast on null input");
}

void test_invalid_create_requests() {
    const StackCreateResultCpp zero_result = stack_create_cpp(0U);
    const StackCreateResultCpp huge_result = stack_create_cpp(std::numeric_limits<std::size_t>::max());
    const StackCreateResultCpp near_overflow_result =
        stack_create_cpp((std::numeric_limits<std::size_t>::max() / sizeof(std::unique_ptr<std::string>)) + 1U);

    record_result(!zero_result.success && zero_result.code == StackStatusCodeCpp::InvalidArgument,
                  "stack_create_cpp rejects zero capacity");
    record_result(!zero_result.message.empty(),
                  "stack_create_cpp provides an error message for zero capacity");
    record_result(!huge_result.success && huge_result.code == StackStatusCodeCpp::InvalidArgument,
                  "stack_create_cpp rejects absurdly large capacity");
    record_result(!near_overflow_result.success && near_overflow_result.code == StackStatusCodeCpp::InvalidArgument,
                  "stack_create_cpp rejects capacity that would overflow item allocation");
}

void test_push_and_pop_lifo() {
    const StackCreateResultCpp create_result = stack_create_cpp(3U);
    StackResultCpp push_result{false, StackStatusCodeCpp::InvalidState, ""};
    StackBoolResultCpp full_result{false, StackStatusCodeCpp::InvalidState, "", false};
    StackBoolResultCpp empty_result{false, StackStatusCodeCpp::InvalidState, "", false};
    StackValueResultCpp top_result{false, StackStatusCodeCpp::InvalidState, "", ""};
    StackValueResultCpp middle_result{false, StackStatusCodeCpp::InvalidState, "", ""};
    StackValueResultCpp bottom_result{false, StackStatusCodeCpp::InvalidState, "", ""};
    StackValueResultCpp underflow_result{false, StackStatusCodeCpp::InvalidState, "", ""};

    record_result(create_result.success && create_result.stack != nullptr,
                  "stack_create_cpp succeeds for LIFO test");
    if (!create_result.success || create_result.stack == nullptr) {
        return;
    }

    push_result = stack_push_cpp(create_result.stack, "first");
    record_result(push_result.success, "stack_push_cpp accepts first element");

    push_result = stack_push_cpp(create_result.stack, "second");
    record_result(push_result.success, "stack_push_cpp accepts second element");

    push_result = stack_push_cpp(create_result.stack, "third");
    record_result(push_result.success, "stack_push_cpp accepts third element");

    full_result = stack_is_full_cpp(create_result.stack);
    record_result(full_result.success && full_result.value,
                  "stack reports full when capacity reached");

    push_result = stack_push_cpp(create_result.stack, "fourth");
    record_result(!push_result.success && push_result.code == StackStatusCodeCpp::Full,
                  "stack_push_cpp rejects overflow");

    top_result = stack_pop_cpp(create_result.stack);
    middle_result = stack_pop_cpp(create_result.stack);
    bottom_result = stack_pop_cpp(create_result.stack);

    record_result(top_result.success && top_result.value == "third",
                  "stack_pop_cpp returns most recent item first");
    record_result(middle_result.success && middle_result.value == "second",
                  "stack_pop_cpp preserves LIFO order");
    record_result(bottom_result.success && bottom_result.value == "first",
                  "stack_pop_cpp returns earliest item last");

    empty_result = stack_is_empty_cpp(create_result.stack);
    record_result(empty_result.success && empty_result.value,
                  "stack is empty after all elements popped");

    underflow_result = stack_pop_cpp(create_result.stack);
    record_result(!underflow_result.success && underflow_result.code == StackStatusCodeCpp::Empty,
                  "stack_pop_cpp returns an empty-stack response on underflow");

    stack_destroy_cpp(create_result.stack);
}

void test_null_argument_handling() {
    StackResultCpp push_result = stack_push_cpp(nullptr, "value");
    StackValueResultCpp pop_result = stack_pop_cpp(nullptr);
    StackBoolResultCpp empty_result = stack_is_empty_cpp(nullptr);
    StackBoolResultCpp full_result = stack_is_full_cpp(nullptr);
    const StackCreateResultCpp create_result = stack_create_cpp(1U);

    record_result(!push_result.success && push_result.code == StackStatusCodeCpp::InvalidArgument,
                  "stack_push_cpp rejects null stack");
    record_result(!push_result.message.empty(),
                  "stack_push_cpp provides an error message for null stack");
    push_result = stack_push_cpp(nullptr, nullptr);
    record_result(!push_result.success && push_result.code == StackStatusCodeCpp::InvalidArgument,
                  "stack_push_cpp rejects null stack and null value");
    record_result(!pop_result.success && pop_result.code == StackStatusCodeCpp::InvalidArgument,
                  "stack_pop_cpp rejects null stack");
    record_result(!empty_result.success && empty_result.code == StackStatusCodeCpp::InvalidArgument,
                  "stack_is_empty_cpp rejects null stack");
    record_result(!full_result.success && full_result.code == StackStatusCodeCpp::InvalidArgument,
                  "stack_is_full_cpp rejects null stack");

    record_result(create_result.success && create_result.stack != nullptr,
                  "stack_create_cpp succeeds for null value test");
    if (create_result.success && create_result.stack != nullptr) {
        push_result = stack_push_cpp(create_result.stack, nullptr);
        record_result(!push_result.success && push_result.code == StackStatusCodeCpp::InvalidArgument,
                      "stack_push_cpp rejects null value on a valid stack");
        record_result(!push_result.message.empty(),
                      "stack_push_cpp provides an error message for null value");
        stack_destroy_cpp(create_result.stack);
    }
}

void test_deep_copy_behavior() {
    const StackCreateResultCpp create_result = stack_create_cpp(2U);
    StackResultCpp push_result{false, StackStatusCodeCpp::InvalidState, ""};
    StackValueResultCpp pop_result{false, StackStatusCodeCpp::InvalidState, "", ""};
    char input[] = "alpha";

    record_result(create_result.success && create_result.stack != nullptr,
                  "stack_create_cpp succeeds for deep-copy test");
    if (!create_result.success || create_result.stack == nullptr) {
        return;
    }

    push_result = stack_push_cpp(create_result.stack, input);
    record_result(push_result.success, "stack_push_cpp accepts mutable caller buffer");
    input[0] = 'X';

    pop_result = stack_pop_cpp(create_result.stack);
    record_result(pop_result.success && !pop_result.value.empty(),
                  "stack_pop_cpp returns stored string after source buffer mutation");
    record_result(pop_result.success && pop_result.value == "alpha",
                  "stack stores a deep copy rather than caller-owned memory");
    record_result(pop_result.success && pop_result.value.c_str() != input,
                  "stack_pop_cpp returns a defensive copy distinct from the caller buffer");

    stack_destroy_cpp(create_result.stack);
}

void test_returned_value_is_defensive_copy() {
    const StackCreateResultCpp create_result = stack_create_cpp(2U);
    StackResultCpp push_result{false, StackStatusCodeCpp::InvalidState, ""};
    StackValueResultCpp first_pop{false, StackStatusCodeCpp::InvalidState, "", ""};
    StackValueResultCpp second_pop{false, StackStatusCodeCpp::InvalidState, "", ""};

    record_result(create_result.success && create_result.stack != nullptr,
                  "stack_create_cpp succeeds for defensive-copy pop test");
    if (!create_result.success || create_result.stack == nullptr) {
        return;
    }

    push_result = stack_push_cpp(create_result.stack, "copy-check");
    record_result(push_result.success, "stack_push_cpp succeeds for defensive-copy pop test");
    if (!push_result.success) {
        stack_destroy_cpp(create_result.stack);
        return;
    }

    first_pop = stack_pop_cpp(create_result.stack);
    record_result(first_pop.success, "first stack_pop_cpp succeeds for defensive-copy pop test");
    if (!first_pop.success) {
        stack_destroy_cpp(create_result.stack);
        return;
    }

    first_pop.value[0] = 'X';
    push_result = stack_push_cpp(create_result.stack, "copy-check");
    record_result(push_result.success, "stack_push_cpp succeeds after caller mutates returned value");

    second_pop = stack_pop_cpp(create_result.stack);
    record_result(second_pop.success && second_pop.value == "copy-check",
                  "caller mutation of returned value does not affect future stack values");

    stack_destroy_cpp(create_result.stack);
}

void test_string_edge_cases() {
    const StackCreateResultCpp create_result = stack_create_cpp(2U);
    StackResultCpp push_result{false, StackStatusCodeCpp::InvalidState, ""};
    StackValueResultCpp long_result{false, StackStatusCodeCpp::InvalidState, "", ""};
    StackValueResultCpp empty_result{false, StackStatusCodeCpp::InvalidState, "", ""};
    char long_input[1024];

    std::memset(long_input, 'A', sizeof(long_input) - 1U);
    long_input[sizeof(long_input) - 1U] = '\0';

    record_result(create_result.success && create_result.stack != nullptr,
                  "stack_create_cpp succeeds for string edge-case test");
    if (!create_result.success || create_result.stack == nullptr) {
        return;
    }

    push_result = stack_push_cpp(create_result.stack, "");
    record_result(push_result.success, "stack_push_cpp accepts empty string");

    push_result = stack_push_cpp(create_result.stack, long_input);
    record_result(push_result.success, "stack_push_cpp accepts long string");

    long_result = stack_pop_cpp(create_result.stack);
    empty_result = stack_pop_cpp(create_result.stack);

    record_result(long_result.success && long_result.value == long_input,
                  "stack_pop_cpp preserves long string contents");
    record_result(empty_result.success && empty_result.value.empty(),
                  "stack_pop_cpp preserves empty string contents");

    stack_destroy_cpp(create_result.stack);
}

void test_destroy_cleans_remaining_items() {
    const StackCreateResultCpp create_result = stack_create_cpp(2U);
    StackResultCpp push_result{false, StackStatusCodeCpp::InvalidState, ""};
    StackResultCpp destroy_result{false, StackStatusCodeCpp::InvalidState, ""};

    record_result(create_result.success && create_result.stack != nullptr,
                  "stack_create_cpp succeeds for destroy cleanup test");
    if (!create_result.success || create_result.stack == nullptr) {
        return;
    }

    push_result = stack_push_cpp(create_result.stack, "secret-1");
    record_result(push_result.success, "stack_push_cpp accepts first cleanup value");

    push_result = stack_push_cpp(create_result.stack, "secret-2");
    record_result(push_result.success, "stack_push_cpp accepts second cleanup value");

    destroy_result = stack_destroy_cpp(create_result.stack);
    record_result(destroy_result.success, "stack_destroy_cpp frees stack with unpopped items");
}

}  // namespace

int main() {
    test_create_and_destroy();
    test_invalid_create_requests();
    test_push_and_pop_lifo();
    test_null_argument_handling();
    test_deep_copy_behavior();
    test_returned_value_is_defensive_copy();
    test_string_edge_cases();
    test_destroy_cleans_remaining_items();

    std::cout << "\nTests run: " << tests_run << '\n';
    std::cout << "Tests failed: " << tests_failed << '\n';

    return tests_failed == 0 ? 0 : 1;
}

File: homework3/java/SecureStackTester.java
import java.util.Arrays;

public final class SecureStackTester {
    private static int testsRun = 0;
    private static int testsFailed = 0;

    private SecureStackTester() {
    }

    public static void main(final String[] args) {
        testCreateAndDestroy();
        testInvalidCreateRequests();
        testPushAndPopLifo();
        testNullArgumentHandling();
        testDeepCopyBehavior();
        testReturnedValueIsDefensiveCopy();
        testStringEdgeCases();
        testDestroyCleansRemainingItems();
        testUseAfterDestroyProtection();

        System.out.println();
        System.out.println("Tests run: " + testsRun);
        System.out.println("Tests failed: " + testsFailed);

        if (testsFailed != 0) {
            System.exit(1);
        }
    }

    private static void recordResult(final boolean condition, final String testName) {
        testsRun++;
        if (condition) {
            System.out.println("[PASS] " + testName);
            return;
        }

        testsFailed++;
        System.out.println("[FAIL] " + testName);
    }

    private static void testCreateAndDestroy() {
        SecureStack.CreateResult createResult = SecureStack.create(3);
        SecureStack.BooleanResult emptyResult;
        SecureStack.BooleanResult fullResult;
        SecureStack.OperationResult destroyResult;

        recordResult(createResult.success() && createResult.stack() != null,
                "create returns a success result and non-null stack");
        if (!createResult.success() || createResult.stack() == null) {
            return;
        }

        emptyResult = createResult.stack().isEmpty();
        fullResult = createResult.stack().isFull();

        recordResult(emptyResult.success() && emptyResult.value(),
                "isEmpty succeeds for a new stack");
        recordResult(fullResult.success() && !fullResult.value(),
                "isFull succeeds for a new stack");

        destroyResult = createResult.stack().destroy();
        recordResult(destroyResult.success(), "destroy succeeds for a valid stack");
    }

    private static void testInvalidCreateRequests() {
        SecureStack.CreateResult zeroResult = SecureStack.create(0);
        SecureStack.CreateResult negativeResult = SecureStack.create(-1);

        recordResult(!zeroResult.success() && zeroResult.code() == SecureStack.StatusCode.INVALID_ARGUMENT,
                "create rejects zero capacity");
        recordResult(zeroResult.message() != null && !zeroResult.message().isEmpty(),
                "create provides an error message for zero capacity");
        recordResult(!negativeResult.success() && negativeResult.code() == SecureStack.StatusCode.INVALID_ARGUMENT,
                "create rejects negative capacity");
    }

    private static void testPushAndPopLifo() {
        SecureStack.CreateResult createResult = SecureStack.create(3);
        SecureStack.OperationResult pushResult;
        SecureStack.BooleanResult fullResult;
        SecureStack.BooleanResult emptyResult;
        SecureStack.ValueResult topResult;
        SecureStack.ValueResult middleResult;
        SecureStack.ValueResult bottomResult;
        SecureStack.ValueResult underflowResult;

        recordResult(createResult.success() && createResult.stack() != null,
                "create succeeds for LIFO test");
        if (!createResult.success() || createResult.stack() == null) {
            return;
        }

        pushResult = createResult.stack().push("first".toCharArray());
        recordResult(pushResult.success(), "push accepts first element");
        pushResult = createResult.stack().push("second".toCharArray());
        recordResult(pushResult.success(), "push accepts second element");
        pushResult = createResult.stack().push("third".toCharArray());
        recordResult(pushResult.success(), "push accepts third element");

        fullResult = createResult.stack().isFull();
        recordResult(fullResult.success() && fullResult.value(),
                "stack reports full when capacity reached");

        pushResult = createResult.stack().push("fourth".toCharArray());
        recordResult(!pushResult.success() && pushResult.code() == SecureStack.StatusCode.FULL,
                "push rejects overflow");

        topResult = createResult.stack().pop();
        middleResult = createResult.stack().pop();
        bottomResult = createResult.stack().pop();

        recordResult(topResult.success() && Arrays.equals(topResult.value(), "third".toCharArray()),
                "pop returns most recent item first");
        recordResult(middleResult.success() && Arrays.equals(middleResult.value(), "second".toCharArray()),
                "pop preserves LIFO order");
        recordResult(bottomResult.success() && Arrays.equals(bottomResult.value(), "first".toCharArray()),
                "pop returns earliest item last");

        emptyResult = createResult.stack().isEmpty();
        recordResult(emptyResult.success() && emptyResult.value(),
                "stack is empty after all elements popped");

        underflowResult = createResult.stack().pop();
        recordResult(!underflowResult.success() && underflowResult.code() == SecureStack.StatusCode.EMPTY,
                "pop returns an empty-stack response on underflow");

        createResult.stack().destroy();
    }

    private static void testNullArgumentHandling() {
        SecureStack.CreateResult createResult = SecureStack.create(1);
        SecureStack.OperationResult pushResult;

        recordResult(createResult.success() && createResult.stack() != null,
                "create succeeds for null input test");
        if (!createResult.success() || createResult.stack() == null) {
            return;
        }

        pushResult = createResult.stack().push(null);
        recordResult(!pushResult.success() && pushResult.code() == SecureStack.StatusCode.INVALID_ARGUMENT,
                "push rejects null input");
        recordResult(pushResult.message() != null && !pushResult.message().isEmpty(),
                "push provides an error message for null input");

        createResult.stack().destroy();
    }

    private static void testDeepCopyBehavior() {
        SecureStack.CreateResult createResult = SecureStack.create(2);
        SecureStack.OperationResult pushResult;
        SecureStack.ValueResult popResult;
        char[] input = "alpha".toCharArray();

        recordResult(createResult.success() && createResult.stack() != null,
                "create succeeds for deep-copy test");
        if (!createResult.success() || createResult.stack() == null) {
            return;
        }

        pushResult = createResult.stack().push(input);
        recordResult(pushResult.success(), "push accepts mutable caller buffer");
        input[0] = 'X';

        popResult = createResult.stack().pop();
        recordResult(popResult.success() && popResult.value() != null,
                "pop returns stored value after source buffer mutation");
        recordResult(popResult.success() && Arrays.equals(popResult.value(), "alpha".toCharArray()),
                "stack stores a deep copy rather than caller-owned memory");
        recordResult(popResult.success() && popResult.value() != input,
                "pop returns a defensive copy distinct from the caller buffer");

        if (popResult.value() != null) {
            Arrays.fill(popResult.value(), '\0');
        }
        createResult.stack().destroy();
    }

    private static void testReturnedValueIsDefensiveCopy() {
        SecureStack.CreateResult createResult = SecureStack.create(2);
        SecureStack.OperationResult pushResult;
        SecureStack.ValueResult firstPop;
        SecureStack.ValueResult secondPop;

        recordResult(createResult.success() && createResult.stack() != null,
                "create succeeds for defensive-copy pop test");
        if (!createResult.success() || createResult.stack() == null) {
            return;
        }

        pushResult = createResult.stack().push("copy-check".toCharArray());
        recordResult(pushResult.success(), "push succeeds for defensive-copy pop test");

        firstPop = createResult.stack().pop();
        recordResult(firstPop.success() && firstPop.value() != null,
                "first pop succeeds for defensive-copy pop test");
        if (!firstPop.success() || firstPop.value() == null) {
            createResult.stack().destroy();
            return;
        }

        firstPop.value()[0] = 'X';
        pushResult = createResult.stack().push("copy-check".toCharArray());
        recordResult(pushResult.success(), "push succeeds after caller mutates returned value");

        secondPop = createResult.stack().pop();
        recordResult(secondPop.success() && Arrays.equals(secondPop.value(), "copy-check".toCharArray()),
                "caller mutation of returned value does not affect future stack values");

        Arrays.fill(firstPop.value(), '\0');
        if (secondPop.value() != null) {
            Arrays.fill(secondPop.value(), '\0');
        }
        createResult.stack().destroy();
    }

    private static void testStringEdgeCases() {
        SecureStack.CreateResult createResult = SecureStack.create(2);
        SecureStack.OperationResult pushResult;
        SecureStack.ValueResult longResult;
        SecureStack.ValueResult emptyResult;
        char[] longInput = new char[1024];

        Arrays.fill(longInput, 'A');

        recordResult(createResult.success() && createResult.stack() != null,
                "create succeeds for string edge-case test");
        if (!createResult.success() || createResult.stack() == null) {
            return;
        }

        pushResult = createResult.stack().push(new char[0]);
        recordResult(pushResult.success(), "push accepts empty char array");
        pushResult = createResult.stack().push(longInput);
        recordResult(pushResult.success(), "push accepts long char array");

        longResult = createResult.stack().pop();
        emptyResult = createResult.stack().pop();

        recordResult(longResult.success() && Arrays.equals(longResult.value(), longInput),
                "pop preserves long array contents");
        recordResult(emptyResult.success() && emptyResult.value() != null && emptyResult.value().length == 0,
                "pop preserves empty array contents");

        if (longResult.value() != null) {
            Arrays.fill(longResult.value(), '\0');
        }
        if (emptyResult.value() != null) {
            Arrays.fill(emptyResult.value(), '\0');
        }
        Arrays.fill(longInput, '\0');
        createResult.stack().destroy();
    }

    private static void testDestroyCleansRemainingItems() {
        SecureStack.CreateResult createResult = SecureStack.create(2);
        SecureStack.OperationResult pushResult;
        SecureStack.OperationResult destroyResult;

        recordResult(createResult.success() && createResult.stack() != null,
                "create succeeds for destroy cleanup test");
        if (!createResult.success() || createResult.stack() == null) {
            return;
        }

        pushResult = createResult.stack().push("secret-1".toCharArray());
        recordResult(pushResult.success(), "push accepts first cleanup value");
        pushResult = createResult.stack().push("secret-2".toCharArray());
        recordResult(pushResult.success(), "push accepts second cleanup value");

        destroyResult = createResult.stack().destroy();
        recordResult(destroyResult.success(), "destroy releases stack with unpopped items");
    }

    private static void testUseAfterDestroyProtection() {
        SecureStack.CreateResult createResult = SecureStack.create(1);
        SecureStack.OperationResult destroyResult;
        SecureStack.OperationResult pushResult;
        SecureStack.ValueResult popResult;
        SecureStack.BooleanResult emptyResult;
        SecureStack.BooleanResult fullResult;

        recordResult(createResult.success() && createResult.stack() != null,
                "create succeeds for use-after-destroy test");
        if (!createResult.success() || createResult.stack() == null) {
            return;
        }

        destroyResult = createResult.stack().destroy();
        recordResult(destroyResult.success(), "destroy succeeds before use-after-destroy checks");

        pushResult = createResult.stack().push("x".toCharArray());
        popResult = createResult.stack().pop();
        emptyResult = createResult.stack().isEmpty();
        fullResult = createResult.stack().isFull();

        recordResult(!pushResult.success() && pushResult.code() == SecureStack.StatusCode.INVALID_STATE,
                "push rejects use after destroy");
        recordResult(!popResult.success() && popResult.code() == SecureStack.StatusCode.INVALID_STATE,
                "pop rejects use after destroy");
        recordResult(!emptyResult.success() && emptyResult.code() == SecureStack.StatusCode.INVALID_STATE,
                "isEmpty rejects use after destroy");
        recordResult(!fullResult.success() && fullResult.code() == SecureStack.StatusCode.INVALID_STATE,
                "isFull rejects use after destroy");
    }
}
